V4L2 driver updates

6by9 opened the pull request:
A few updates to the V4L2 driver for the Pi Camera.
- Correct the support for ISO control.
- Add option to specify the /dev/videoN node number.
- Support for multiple cameras on the CM.
- Support for specifying where the overlay is on the screen.
- Compliance test failure.

A quick review by others would be appreciated, just to make sure I haven't missed something silly in the init/exit loops. (I'm afraid I haven't done as much testing of that as I perhaps ought to have done)

6by9 committed:
BCM2835-V4L2: Correct ISO control and add V4L2_CID_ISO_SENSITIVITY_AUTO

https://github.com/raspberrypi/linux/issues/1251

V4L2_CID_ISO_SENSITIVITY was not advertising ISO*1000 as it should.
V4L2_CID_ISO_SENSITIVITY_AUTO was not implemented, so was taking
V4L2_CID_ISO_SENSITIVITY as 0 for auto mode.
Still accepts 0 for auto, but also abides by the new parameter.

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

6by9 committed:
BCM2835-V4L2: Add a video_nr parameter.

Adds a kernel parameter "video_nr" to specify the preferred
/dev/videoX device node.
https://www.raspberrypi.org/forums/viewtopic.php?f=38&t=136120&p=905545

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

6by9 committed:
BCM2835-V4L2: Add control of the overlay location and alpha.

Actually do something useful in vidioc_s_fmt_vid_overlay and
vidioc_try_fmt_vid_overlay, rather than effectively having
read-only fields.

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

6by9 committed:
BCM2835-V4L2: V4L2-Compliance failure fix

VIDIOC_TRY_FMT was failing due to bytesperline not
being set correctly by default.

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

popcornmix commented on drivers/media/platform/bcm2835/bcm2835-camera.c:
```diff
@@ -54,6 +54,10 @@ int bcm2835_v4l2_debug;
 module_param_named(debug, bcm2835_v4l2_debug, int, 0644);
 MODULE_PARM_DESC(bcm2835_v4l2_debug, "Debug level 0-2");
 
+int video_nr = -1;
```
static?

6by9 replied:
Perhaps same argument should be applied to bcm2835_v4l2_debug above.  As you've seen, I've copy/pasted from somewhere else in the next patch and it has become static :-)

popcornmix commented on drivers/media/platform/bcm2835/bcm2835-camera.c:
```diff
@@ -54,9 +56,10 @@ int bcm2835_v4l2_debug;
-module_param_named(video_nr, video_nr, int, 0644);
-MODULE_PARM_DESC(video_nr, "videoX start number, -1 is autodetect");
+#define UNSET (-1)
+static int video_nr[] = {[0 ... (MAX_BCM2835_CAMERAS - 1)] = UNSET };
```
Ah - static!

popcornmix commented on drivers/media/platform/bcm2835/bcm2835-camera.c:
```diff
@@ -1792,8 +1895,10 @@ static int __init bm2835_mmal_init(void)
-	v4l2_err(&dev->v4l2_dev,
-		 "%s: error %d while loading driver\n",
+	for ( ; camera > 0; camera--)
+		bcm2835_cleanup_instance(gdev[camera]);
```
Should this cleanup gdev[0]?
Should you set gdev[camera] = NULL; ?

6by9 replied:
Possibly. 
This is failure to load, so won't that unload the module due to the failure and therefore deallocate gdev.

Probably safest to reset it to NULL anyway - I'll update it.

6by9 committed:
BCM2835-V4L2: Make all module parameters static

Clean up to correct variable scope

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

6by9 committed:
BCM2835-V4L2: Add support for multiple cameras

Ask GPU on load how many cameras have been detected, and
enumerate that number of devices.
Only applicable on the Compute Module as no other device
exposes multiple CSI2 interfaces.

Signed-off-by: Dave Stevenson <6by9@users.noreply.github.com>

6by9 commented:
@popcornmix I think that addresses the things you raised, and adds an extra patch making the module parameters static variables, except bcm2835_v4l2_debug which is used in multiple files).
